File: SumRootToLeafNumbers.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

def representNumber(node):
    if node == None:
        return []
    leftNumbers = []
    rightNumbers = []
    if node.left != None:
        leftNumbers = representNumber(node.left)
    if node.right != None:
        rightNumbers = representNumber(node.right)
    if len(leftNumbers) == 0 and len(rightNumbers) == 0:
        return [node.val]
    result = []
    for leftnumber in leftNumbers:
        result.append(str(node.val) + str(leftnumber))
    for rightnumber in rightNumbers:
        result.append(str(node.val) + str(rightnumber))
    return result

class Solution:
    # @param root, a tree node
    # @return an integer
    def sumNumbers(self, root):
        if root == None:
            return 0
        result = 0
        numbers = representNumber(root)
        for number in numbers:
            result += int(number)
        return result

File: SumRootToLeafNumbers.cc
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode *root) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        int result = 0;
        helper(root, &result);
        return result;
    }
    
    void helper(TreeNode * root, int * current)
    {
        if(!root){
            //*current = 0;
            return;
        }
        *current = *current * 10 + root->val;
        int left = *current, right = *current;
        if(root->left)
            helper(root->left, &left);
        else
            left = 0;
        if(root->right)
            helper(root->right, &right);
        else
            right = 0;
        if(left != 0 || right != 0)
            *current = left + right;
    }


};
